tasks/release/ReleaseVersionController.php

```diff
@@ -36,15 +36,50 @@ function assignVersionToCompletedTickets($version = null) {
     return [$version, join("\n", $output)];
   }
 
-  private function getDoneIssuesWithoutVersion() {
-    $jql = "project = $this->project AND status = Done AND fixVersion = EMPTY AND updated >= -52w";
-    $result = $this->jira->search($jql, ['key']);
-    return array_map(function ($issue) {
-      return [
-        'id' => $issue['id'],
-        'key' => $issue['key'],
-      ];
-    }, $result['issues']);
+  function determineNextVersion() {
+    $last_version = $this->jira->getLastReleasedVersion();
+    $parsed_last_version = VersionHelper::parseVersion($last_version['name']);
+
+    $version_to_increment = VersionHelper::PATCH;
+
+    if ($this->project === JiraController::PROJECT_MAILPOET) {
+      $free_increment = $this->checkProjectVersionIncrement(JiraController::PROJECT_MAILPOET);
+      $premium_increment = $this->checkProjectVersionIncrement(JiraController::PROJECT_PREMIUM);
+
+      if (in_array(VersionHelper::MINOR, [$free_increment, $premium_increment])) {
+        $version_to_increment = VersionHelper::MINOR;
+      }
+    }
+
+    $parsed_last_version[$version_to_increment]++;
+    $next_version = VersionHelper::buildVersion($parsed_last_version);
+    return $next_version;
+  }
+
+  private function checkProjectVersionIncrement($project) {
+    $this->jira->setProject($project);
+    $issues = $this->getUnreleasedDoneIssues();
+    $this->jira->setProject(JiraController::PROJECT_MAILPOET); // restore project
+
+    $version_to_increment = VersionHelper::PATCH;
+    $field_id = JiraController::VERSION_INCREMENT_FIELD_ID;
+
+    foreach ($issues as $issue) {
+      if (!empty($issue['fields'][$field_id]['value'])
+        && $issue['fields'][$field_id]['value'] === VersionHelper::MINOR
+      ) {
+        $version_to_increment = VersionHelper::MINOR;
+        break;
+      }
+    }
+
+    return $version_to_increment;
+  }
+
+  private function getUnreleasedDoneIssues() {
+    $jql = "project = $this->project AND status = Done AND (fixVersion = EMPTY OR fixVersion IN unreleasedVersions()) AND updated >= -52w";
```

Comment:
Two questions here:
1. Why do we need the `updated >= -52w`? Some perf optimization?
2. Do we want to pick all unreleased tickets or do we actually first want to assign the new release version (another step in our release checklist) and then here check only against that version? It's just a question, I don't know which approach is better.

Reply:
1. I took this from the Jira interface's queries, and it seems to be necessary, otherwise it picks up some odd ancient tickets.
2. See item no. 3 from specs, we want to pick all unreleased tickets.